src/mlpack/methods/lmnn/lmnn_main.cpp

```diff
@@ -0,0 +1,468 @@
+    "\n\n" +
+    PRINT_CALL("mlpack_lmnn", "input", "letter_recognition", "k", 5,
+    "range", 10, "regularization", 0.4, "output", "output"));
+
+PARAM_MATRIX_IN_REQ("input", "Input dataset to run LMNN on.", "i");
+PARAM_MATRIX_IN("distance", "Initial distance matrix to be used as "
+    "starting point", "d");
+PARAM_UROW_IN("labels", "Labels for input dataset.", "l");
+PARAM_INT_IN("k", "Number of target neighbors to use for each "
+    "datapoint.", "k", 1);
+PARAM_MATRIX_OUT("output", "Output matrix for learned distance matrix.", "o");
+PARAM_MATRIX_OUT("transformed_data", "Output matrix for transformed dataset.",
+    "D");
+PARAM_FLAG("print_accuracy", "Print accuracies on initial and transformed "
+    "dataset", "P");
+PARAM_STRING_IN("optimizer", "Optimizer to use; 'amsgrad', 'bbsgd', 'sgd', or "
+    "'lbfgs'.", "O", "amsgrad");
+PARAM_DOUBLE_IN("regularization", "Regularization for LMNN objective function ",
+    "r", 0.5);
+PARAM_FLAG("normalize", "Use a normalized starting point for optimization. It"
+    "is useful for when points are far apart, or when SGD is returning NaN.",
+    "N");
+PARAM_INT_IN("passes", "Maximum number of full passes over dataset for "
+    "AMSGrad, BB_SGD and SGD.", "p", 50);
+PARAM_INT_IN("max_iterations", "Maximum number of iterations for "
+    "L-BFGS (0 indicates no limit).", "n", 100000);
+PARAM_DOUBLE_IN("tolerance", "Maximum tolerance for termination of AMSGrad, "
+    "BB_SGD, SGD or L-BFGS.", "t", 1e-7);
+PARAM_DOUBLE_IN("step_size", "Step size for AMSGrad, BB_SGD and SGD (alpha).",
+    "a", 0.01);
+PARAM_FLAG("linear_scan", "Don't shuffle the order in which data points are "
+    "visited for SGD or mini-batch SGD.", "L");
+PARAM_INT_IN("batch_size", "Batch size for mini-batch SGD.", "b", 50);
+
+PARAM_DOUBLE_IN("beta1", "Exponential decay rate for the first moment "
+    "estimates of AMSGrad.", "x", 0.9);
+PARAM_DOUBLE_IN("beta2", "Exponential decay rate for weighted infinity norm "
+    "estimates of AMSGrad.", "y", 0.999);
+PARAM_DOUBLE_IN("epsilon", "Value used to initialise the mean squared gradient "
+    "parameter of AMSGrad.", "e", 1e-8);
+PARAM_DOUBLE_IN("batch_delta", "Factor for the batch update step of "
+  "BigBatch_SGD.", "E", 0.1);
+PARAM_INT_IN("num_basis", "Number of memory points to be stored for L-BFGS.",
+    "B", 5);
+PARAM_DOUBLE_IN("armijo_constant", "Armijo constant for L-BFGS.", "A", 1e-4);
+PARAM_DOUBLE_IN("wolfe", "Wolfe condition parameter for L-BFGS.", "w", 0.9);
+PARAM_INT_IN("max_line_search_trials", "Maximum number of line search trials "
+    "for L-BFGS.", "T", 50);
+PARAM_DOUBLE_IN("min_step", "Minimum step of line search for L-BFGS.", "m",
+    1e-20);
+PARAM_DOUBLE_IN("max_step", "Maximum step of line search for L-BFGS.", "M",
+    1e20);
+PARAM_INT_IN("range", "Number of iterations after which impostors needs to be "
+    "recalculated", "R", 1);
+PARAM_INT_IN("seed", "Random seed.  If 0, 'std::time(NULL)' is used.", "s", 0);
```

Comment:
Right, this eventually got so big, that I had a hard time finding alphabets for them :)
It would be good enough if we can just avoid all basic optimizers specific parameters, I guess there are many like beta1, beta2, etc. Probably they will be used occasionally.

Reply:
Ok---if you want to simplify before merge, feel free; otherwise we can open an issue.

Reply:
Done!

Reply:
I think it's fine to remove `num_basis`, `armijo_constant`, and `wolfe`, if you like, since these don't usually need to be tuned for L-BFGS either (in fact L-BFGS doesn't tend to need tuning much at all).